docs/proposals/spec-reconciliation.md

```diff
@@ -0,0 +1,130 @@
+
+#### Approach#2
+Error out the provision request. It is clear with the polling approach of the Service Catalog to update the [specs](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#response-2), that this scenario will eventually popup, but the specs doesnt have an error code specifically meant for this scenario.
+
+#### Approach#3
+It can send a 202 Accepted status, suggesting that provisioning is in progress. The service catalog will when keep polling to check if the service instance is fully provisioned. The [last operation](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#response-1) has a mechanism where broker can tell if the provision is success of failure.
+
+
+## Other ideas
+
+These ideas are not directly related to the objectives of this proposal but are mentioned here and might become independent proposals of their own when addressing those issues.
+
+#### APB Versioning
+
+Currently, tags are the only way an apb developer can version an apb. If the developer is given a way to do an explicit versioning of the apb irrespective of the tags, there might never be a need to allow multiple tags
+
+A rough idea of what can be done:
+
+ 1. Add a new field to apb.yml call bundleVersion
+ 2. During provision save the value of "bundleVersion" with the service instance
+ 3. During de-provision pass the value of "bundleVersion" in the args.
+ 
+This allows:
+ 1. the developer to have a versioning scheme that is independent of the tags.
+ 2. the developer can push v1.1 of that apb even if v1.0 is provisioned. The version number passed during binding, unbinding and deprovision would still be v1.0
+ 3. the admin can know what version of apb is provisioned in the cluster
+ 
+
+#### Multiple tags
+
+The requirement supporting multiple tags of apb is largely because versioning of an apb is tied to the tag right now. There needs to be a way in which admins can have multiple versions of the same apb. If the apb versioning scheme mentioned above is implemented, there might not be a need to support multiple tags
+
+If there is, the registry section of configmap of the broker can be modified as follows:
+
+    registry:
+      - type: dockerhub
+        name: dh
+        url: https://registry.hub.docker.com
+        org: ansibleplaybookbundle
+        tag:
+          default: latest
+          named: 
+            - hello-apb: 
+              - v1.0
+              - v2.0
+            - postgres-apb: 
+              - v3.9
+        white_list:
+          - '.*-apb$'
+
+
+This would require modification to the bootstrap function where after fetching all the specs, the broker needs to iterate through the list of named specs and fetch specific tags from the registry
+
+Adding a method to adaptor interface that takes two inputs, image name and image tag, and fetches its spec. The broker will then call this method iteratively for all named pair of specs and tags
```

Comment:
@jmrodri @eriknelson @shawn-hurley I think the ***in-place merge*** (approach 1) could still be a good way to achieve reconciliation. My modified approach is as follows:

1. Maintain a ***bundle instance counter*** along with each spec in the CRD.
2. Compare the list of registrySpecs with crdSpecs.
3. Add all the specs that present in registrySpecList but absent in crdSpecList, with bundle instance counter initialized to 0
4. Remove all specs present in crdSpecList but absent in registrySpecList and with bundle instance counter value 0

Increment the spec bundle instance counter during provision
Decrement the spec bundle instance counter during deprovision

I think step 2 and 3 in the above algorithm would be the ones that would take most of the time. But those steps would be required even if the ***reaper approach*** is taken, to mark the specs for deletion. Maintaining the counter value would help save the time consuming computation required to determine whether a spec has a bundle instance. This will also avoid all the concurrency problems that complicates the ***reaper approach***

Reply:
I don't like maintaining a separate counter because that is information we can calculate by looking for instances that have references to the specs. While that's not trivial, it is accurate. My gut tells me those counters are going to get out of date and create another issue. But that's just me, I'll let others chime in

Reply:
@jmrodri @eriknelson @shawn-hurley I think modifying the reaper approach as follows could be one way to handle reconciliation with some simplicity:

1. There is a background process reaper which listens for data from bootstrap. Bootstrap runs periodically as it does now.
2. The bootstrap process is responsible for determining which spec should be added and adding them.
3. The bootstrap process will then determine which spec to be deleted and send them to reaper(one by one as it determines).
4. The reaper always runs in the background waiting to get a spec from bootstrap. When the reaper gets a spec, it deletes the spec if there are no bundle instances, else it will ignore the spec.

This simplified approach gets rid of the **mark for deletion** field. The reaper now tries to **safely delete** the specs it gets from bootstrap. If it is not safely able to delete, it will ignore it. Since the adding of specs is done by bootstrap, and safe removal is done by reaper, they no longer compete for a shared resource. This will always ensure a desirable state of the broker.

Reply:
Unfortunately the delete is not the expensive operation, it's the iteration of outstanding service instances to determine which instances correspond to specs, so it's not desirable to have the bootstrap process determine the specs that are safe to delete. The purpose of the reaper is to delegate this work to the background.